### PR TITLE
Tin 17

### PR DESCRIPTION
**Pull request by a40o**
Resend verification email

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -44,7 +44,7 @@ dependencies {
     implementation 'org.passay:passay:1.6.1'
     implementation 'org.springdoc:springdoc-openapi-starter-webmvc-ui:2.1.0'
     implementation 'org.springframework.boot:spring-boot-starter-mail'
-    implementation 'com.google.googlejavaformat:google-java-format:$googleJavaFormatVersion'
+//    implementation 'com.google.googlejavaformat:google-java-format:$googleJavaFormatVersion'
```
Comment (todornikolov1) [inline build.gradle]: Why this is commented? No need to keep it, if we don't want to use it

File: src/main/java/com/volasoftware/tinder/services/UserService.java [hunk 2]
```diff
@@ -78,12 +72,11 @@ public void registerUser(UserDto userDto) throws MessagingException, IOException
...
 
-        MimeMessage message = mailSender.createMimeMessage();
-        message.setFrom(new InternetAddress("ang3lkirilov@gmail.com"));
-        message.setRecipients(MimeMessage.RecipientType.TO,user.getEmail());
-        message.setSubject("Verification");
-        message.setContent(getEmailContent(token.getToken()),"text/html; charset=utf-8");
-        mailSender.send(message);
+        emailSenderService.sendEmail(user,
+                token,
+                "ang3lkirilov@gmail.com",
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/UserService.java]: make this a property into application.properties

File: src/main/java/com/volasoftware/tinder/services/VerificationService.java [hunk 3]
```diff
@@ -38,4 +48,32 @@ public boolean verifyUser(String token){
...
+                    () -> new UserDoesNotExistException("This user does not exist!")
+            );
+
+            if (user.isEnabled()){
+                throw new UserAlreadyVerifiedException("User is already verified!");
+            }
+
+            Optional<Verification> oldToken = verificationRepository.findTokenByUserId(user.getId());
+
+            oldToken.ifPresent(verificationRepository::delete);
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/VerificationService.java]:
```suggestion
            verificationRepository.findTokenByUserId(user.getId()).ifPresent(verificationRepository::delete);
```

File: src/main/java/com/volasoftware/tinder/services/UserService.java [hunk 4]
```diff
@@ -78,12 +72,11 @@ public void registerUser(UserDto userDto) throws MessagingException, IOException
...
+                "Verification",
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/UserService.java]: create a constant and use it

File: src/main/java/com/volasoftware/tinder/services/UserService.java [hunk 5]
```diff
@@ -78,12 +72,11 @@ public void registerUser(UserDto userDto) throws MessagingException, IOException
...
+                "text/html; charset=utf-8");
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/UserService.java]: another constant candidate

File: src/main/java/com/volasoftware/tinder/services/VerificationService.java [hunk 6]
```diff
@@ -38,4 +48,32 @@ public boolean verifyUser(String token){
...
+            newToken.setToken(UUID.randomUUID().toString());
+            newToken.setCreatedDate(LocalDateTime.now());
+            newToken.setExpirationDate(LocalDateTime.now().plusDays(2));
+            verificationRepository.saveAndFlush(newToken);
+
+            emailSenderService.sendEmail(user,
+                    newToken,
+                    "ang3lkirilov@gmail.com",
+                    "Verification",
+                    "text/html; charset=utf-8");
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/VerificationService.java]: use defined properties and constants instead

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    private String getEmailContent(String token) throws IOException{
+
+        Resource emailResource = resourceLoader.getResource("classpath:emailResources/ConfirmationPage.html");
+
+        File emailFile = emailResource.getFile();
+        Path path = Path.of(emailFile.getPath());
+        String emailContent = Files.readString(path);
+
+        return  emailContent.replace("{{token}}" , "http://localhost:8080/verify/" + token);
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: `http://localhost:8080/` is base URL with is environment related, should be in application.properties

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        message.setFrom(new InternetAddress(sender));
+        message.setRecipients(MimeMessage.RecipientType.TO,user.getEmail());
+        message.setSubject(subject);
+        message.setContent(getEmailContent(token.getToken()),contentType);
+        mailSender.send(message);
+    }
+
+    private String getEmailContent(String token) throws IOException{
+
+        Resource emailResource = resourceLoader.getResource("classpath:emailResources/ConfirmationPage.html");
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: Probable is better to not depend on a resource here. All logic about getting an content that will be send to email (or whatever) is user/token domain specific, no any real connection about sending email.

My advice is to move somewhere else the whole logic and to add one parameter more as content (just an example name) to sendEmail() above

**Comment by todornikolov1** [comment]
Any plans to add tests to cover the logic as part of this PR?

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+@Service
+public class EmailSenderService {
+
+    private JavaMailSender mailSender;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: private final?

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Value("{$verification_link}")
+    private String verificationLink;
+
+    public EmailSenderService(JavaMailSender mailSender, ResourceLoader resourceLoader) {
+        this.mailSender = mailSender;
+        this.resourceLoader = resourceLoader;
+    }
+
+    public void sendEmail(User user,
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: inject the subject and the content as params

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                          Verification token) throws MessagingException, IOException {
+        MimeMessage message = mailSender.createMimeMessage();
+        message.setFrom(new InternetAddress(emailSender));
+        message.setRecipients(MimeMessage.RecipientType.TO,user.getEmail());
+        message.setSubject(emailSubject);
+        message.setContent(getEmailContent(token.getToken()),emailContentType);
+        mailSender.send(message);
+    }
+
+    private String getEmailContent(String token) throws IOException{
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: Move to a service related with the token and just call emailSenderService.sendEmail(email, subject, content) from there

File: src/test/java/com/volasoftware/tinder/UserRepositoryTest.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    public void testCreateUser(){
+
+        User user = new User();
+        user.setEmail("test@testmail.com");
+        user.setPassword("11!!TestTest");
+        user.setFirstName("FirstName");
+        user.setLastName("LastName");
+        user.setGender(Gender.MALE);
+
+        User savedUser = userRepository.save(user);
```
Comment (todornikolov1) [inline src/test/java/com/volasoftware/tinder/UserRepositoryTest.java:36]: check user existence in db after the test passed successfully

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java (hunk 10 above)
Comment (a40o) [reply src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: By "content" you mean all after ".setContent" or just "emailContentType"?

File: src/main/java/com/volasoftware/tinder/controller/VerificationController.java [hunk 13]
```diff
@@ -1,22 +1,57 @@
...
+
 @RestController
 public class VerificationController {
 
     private final VerificationService verificationService;
+    private final VerificationRepository verificationRepository;
+    private final JavaMailSender mailSender;
+    private final UserService userService;
+    private final UserRepository userRepository;
+    private final EmailSenderService emailSenderService;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/controller/VerificationController.java]: Too many added dependences. I think the only one used is verificationService.

File: src/main/java/com/volasoftware/tinder/exception/UserAlreadyVerifiedException.java [hunk 14]
```diff
@@ -0,0 +1,6 @@
+package com.volasoftware.tinder.exception;
+public class UserAlreadyVerifiedException extends RuntimeException{
+    public UserAlreadyVerifiedException(String message){
+        super(message);
+    }
+}
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/exception/UserAlreadyVerifiedException.java:7]: format the code (Ctrl+Alt+L)

File: src/main/java/com/volasoftware/tinder/exception/UserDoesNotExistException.java [hunk 15]
```diff
@@ -0,0 +1,5 @@
+package com.volasoftware.tinder.exception;public class UserDoesNotExistException extends RuntimeException{
+    public UserDoesNotExistException(String message){
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/exception/UserDoesNotExistException.java]: use Ctrl+Alt+L

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.io.IOException;
+import java.util.Set;
+
+@Service
+public class EmailSenderService {
+
+    private final JavaMailSender mailSender;
+    @Value("${spring.mail.username}")
+    private String emailSender;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]:
```suggestion
    private final String emailSender;
```

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    private final UserRepository userRepository;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: Not used, remove it

File: src/main/java/com/volasoftware/tinder/services/UserService.java [hunk 18]
```diff
@@ -32,30 +24,28 @@ public class UserService {
     private final VerificationRepository verificationRepository;
     private final ResourceLoader resourceLoader;
     private JavaMailSender mailSender;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/UserService.java]: Do we use that?

Comment (a40o) [reply src/main/java/com/volasoftware/tinder/services/UserService.java]: I've already removed the 26th and 25th line, but I'm using the verification repository to save the token on line 62

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java (hunk 10 above)
Comment (todornikolov1) [reply src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: I mean, same as the below comment. To add a parameter content of type String, to pass the whole email body text with it

File: src/main/java/com/volasoftware/tinder/services/VerificationService.java [hunk 19]
```diff
@@ -38,4 +58,38 @@ public boolean verifyUser(String token){
...
+            Verification newToken = new Verification();
+            newToken.setUser(user);
+            newToken.setToken(UUID.randomUUID().toString());
+            newToken.setCreatedDate(LocalDateTime.now());
+            newToken.setExpirationDate(LocalDateTime.now().plusDays(2));
+            verificationRepository.saveAndFlush(newToken);
+
+            emailSenderService.sendEmail("Verification",
+                    Collections.singleton(user.getEmail()),
+                    setVerificationLink(newToken.getToken()));}
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/VerificationService.java]: Move { to next line

File: src/main/java/com/volasoftware/tinder/controller/VerificationController.java [hunk 20]
```diff
@@ -1,16 +1,22 @@
...
 import org.springframework.web.bind.annotation.RestController;
 
+import java.io.IOException;
+
 @RestController
 public class VerificationController {
 
     private final VerificationService verificationService;
 
+
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/controller/VerificationController.java]: No need of this extra line

File: src/main/java/com/volasoftware/tinder/services/VerificationService.java [hunk 21]
```diff
@@ -38,4 +58,39 @@ public boolean verifyUser(String token){
...
+            public String setVerificationLink(String token) throws IOException{
+
+                Resource emailResource = resourceLoader.getResource("classpath:emailResources/ConfirmationPage.html");
+
+                File emailFile = emailResource.getFile();
+                Path path = Path.of(emailFile.getPath());
+                String emailContent = Files.readString(path);
+
+                return  emailContent.replace("{{token}}" , hostLink + "verify/" + token);
+            }
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/VerificationService.java]: Added one extra tab of indentation to the whole method?

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java (hunk 9 above)
Comment (a40o) [reply src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: Added

File: src/main/java/com/volasoftware/tinder/services/EmailSenderService.java (hunk 11 above)
Comment (a40o) [reply src/main/java/com/volasoftware/tinder/services/EmailSenderService.java]: Done

File: src/test/java/com/volasoftware/tinder/UserRepositoryTest.java (hunk 12 above)
Comment (a40o) [reply src/test/java/com/volasoftware/tinder/UserRepositoryTest.java:36]: @todornikolov1 this will be resolved as part of the next task tests.

**Pull request merged**